File: src/fluxharmonium/dacs.py
"""
fluxharmonium dac value generator

processes all values to determine and scale the DAC values
and write them to the serial bus
"""
import asyncio
from itertools import chain

import math

from config import *

def clamp(n, min, max):
    if n < min: 
        return min
    elif n > max: 
        return max
    else: 
        return n 


class DACs():
    """
    Setup a DAC control loop to output the dac values
    """
    def __init__(self, parent, frequency, env_vals, vel_vals):
        self.parent = parent

        self.frequency = frequency
            
        self.env_vals = env_vals
        self.vel_vals = vel_vals
        
        self.dac_max_value = 255
        dsv = self.dac_max_value
        
        self.dac_values = [
            [dsv, dsv, dsv, dsv, dsv, dsv, dsv, dsv],
            [dsv, dsv, dsv, dsv, dsv, dsv, dsv, dsv],
            [dsv, dsv, dsv, dsv, dsv, dsv, dsv, dsv],
            [dsv, dsv, dsv, dsv, dsv, dsv, dsv, dsv],
            [dsv, dsv, dsv, dsv, dsv, dsv, dsv, dsv],
            [dsv, dsv, dsv, dsv, dsv, dsv, dsv, dsv],
            [dsv, dsv, dsv, dsv, dsv, dsv, dsv, dsv],
            [dsv, dsv, dsv, dsv, dsv, dsv, dsv, dsv],
            [dsv, dsv, dsv, dsv, dsv, dsv, dsv, dsv],
            [dsv, dsv, dsv, dsv, dsv, dsv, dsv, dsv],
            [dsv, dsv, dsv, dsv, dsv, dsv, dsv, dsv],
            [dsv, dsv, dsv, dsv, dsv, dsv, dsv, dsv],
        ]
        
        self.group_size = len(self.dac_values[0])
        
        self.parent._logger.debug('sending intial dac_values')
        self.parent.send_dac_values(self.dac_values)

    def freq_to_sec(self, freq):
        return 1000/1000/freq
    
    def get_group_index(self, position):
        output = position
        
        group = int(((output-1) / self.group_size))
        index = int(((output-1) % self.group_size))
        
        # self.parent._logger.debug(f'output: {output} group: {group} index: {index}')
        
        return {'group': group, 'index': index}
    
    def invert(self, value):
        return self.dac_max_value - value

    def scale_midi_to_dac(self, x):
        # (val - in_min) * (out_max - out_min) / (in_max - in_min) + out_min
        MIDI_MAX = 127
        return int(x * (DAC_MAX_RANGE - DAC_MIN_RANGE) / MIDI_MAX + DAC_MIN_RANGE) 

    
    def calculate_dac_values(self):
        """
        Missing the mapping to the output range
        """
        for dac in range(NUMBER_OF_NOTES):
            dac_index = self.get_group_index(NOTE_MAP[dac+MIDI_LOW_VALUE]-1)
            try:
                current_value= clamp(self.invert(
                        self.scale_midi_to_dac(
                            math.floor(self.env_vals[dac] * self.vel_vals[dac])
                        )
                    ),0, 255)

                self.dac_values[dac_index['group']][dac_index['index']] = current_value
                # if dac == 0:
                #     self.parent._logger.info(f'dac 0 {current_value}')
            except Exception as e:
                self.parent._logger.error(f'Could not calculate the dac value {dac} {e}')

    async def run(self):
        while True:
            # self.parent._logger.debug('dac cycle')
            self.calculate_dac_values()
            self.parent.send_dac_values(self.dac_values)
            await asyncio.sleep(self.freq_to_sec(self.frequency))

File: src/teensy/arduino/src/dacs.cpp
#include "dacs.h"

// DAC Constructor
DACs::DACs(float* env_vals, uint8_t* vel_vals, bool debug) {  
      DACs::m_env_vals = env_vals;
      DACs::m_vel_vals = vel_vals;
      DACs::m_debug = debug;
}

void DACs::init() {
  // Allocate memory for the 2D dac_values array
  m_dac_values = new uint16_t [NUMBER_OF_NOTES];
  memset(m_dac_values, 255, 96 * sizeof(uint8_t));

  m_dac_values[0] = 0;

  m_ltc1665.init(uint8_t(10), m_debug);
  m_ltc1665.write_chained_dac_values(m_dac_values, NUMBER_OF_NOTES);

  if (m_debug) {
      Serial.println("DACs init'd");
      printDacs();
  }

}

// Destructor with memory deallocation for dac_values
// DACs::~DACs() {
//   for (int i = 0; i < NUMBER_OF_NOTES; ++i) {
//     delete[] dac_values[i];
//   }
//   delete[] dac_values;
// }

// Invert DAC value (assuming higher value means lower output)
uint8_t DACs::invert_dac_value(uint8_t value) {
  return uint8_t(DAC_MAX_RANGE - value);
}

// Scale MIDI value to DAC range
uint8_t DACs::scale_midi_to_dac(uint8_t x) {
  const uint8_t MIDI_MAX = 127;
  return uint8_t(x * (DAC_MAX_RANGE - DAC_MIN_RANGE) / MIDI_MAX + DAC_MIN_RANGE);
}

// Calculate DAC values for all channels
void DACs::calculate_dac_values() {
  for (int dac = 0; dac < NUMBER_OF_NOTES; dac++) {
    uint8_t position = NOTE_MAP[(dac + MIDI_LOW_VALUE)]-1;

    m_dac_values[position] = invert_dac_value(
        scale_midi_to_dac(uint8_t((m_env_vals[dac] * m_vel_vals[dac])))
    );

    // Optional debug print for a specific note
    // if (dac + MIDI_LOW_VALUE == 60) {
    //   Serial.printf("Note 60 env_val: %d vel_val: %d dac_val: %d", m_env_vals[dac], m_vel_vals[dac], m_dac_values[group_index][channel_index]);
    // }
  }
  Serial.println("Calculated dacs");
}

// Function to run in the DAC loop
void DACs::run() {
    calculate_dac_values();
    if (m_debug) {
      printDacs();
    }
    m_ltc1665.write_chained_dac_values(m_dac_values, NUMBER_OF_NOTES);
    Serial.println("DAC running");
}

void DACs::printDacs() {
  Serial.println("dac_values: ");
    for (int i=0; i<NUMBER_OF_NOTES;i++){
      if (i % DAC_GROUP == 0){
        Serial.print("\n[");
      }
      
      Serial.print(m_dac_values[i]);
      
      if (i % DAC_GROUP == DAC_GROUP-1){
        Serial.println("]");
      } else {
        Serial.print(",");
      }
    }
    Serial.println("");
    Serial.println("");
}
